### PR TITLE
refactor(iter): use circle_buff to simplify codes

### PR DESCRIPTION
**Pull request by W-Mai**
- Add new `lv_circle_buff_t` data structure to easily and efficiently handle cases when using queue or ping-pong buffer and so on.
- Simplify `lv_iter_t`. Use `lv_circle_buff_t` to simplify the iter peek and next logics.

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` (`astyle v3.4.12` needs to installed by running `cd scripts; ./install_astyle.sh`) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

**Comment by W-Mai** [comment]
> Isn't it a FIFO?

circle fifo I think 😁

**Comment by kisvegabor** [comment]
What about calling it FIFO then? :slightly_smiling_face:

**Comment by W-Mai** [comment]
> What about calling it FIFO then? 🙂

The same structure in the Nuttx OS name circle buff, but I think fifo is OK.

I think the diff from fifo is circle buff has an ability of random accessing.

What do you think?

**Comment by kisvegabor** [comment]
> I think the diff from fifo is circle buff has an ability of random accessing.

That's really an important difference.  Let's keep the current name.

---

I wonder if it's really worth adding a circular buffer though. By looking in `lv_iter.c` I can't see the it's significantly shorter, but a whole new module was added. Maybe I missing where the circular buffer help that much. Could you point it out?  If it's really not a game changer I rather not add it for a single use case.

**Comment by W-Mai** [comment]
It is not to make the code look more concise from the number of lines, but to make the logic clearer.

The previous iter's peek buffer requires extra memmove to move data. After using circle buff, there is no need to move data, and the structure of iter is also simpler.

In addition, if you want to add infinite scrolling components, multi-buffer management, etc. in lvgl in the future, using circle buffer is also the best choice, I think.

**Comment by kisvegabor** [comment]
> In addition, if you want to add infinite scrolling components, multi-buffer management, etc. in lvgl in the future, using circle buffer is also the best choice, I think.

You have just sold it to me with the hope for infinite scrolling 😀

**Review by kisvegabor — DISMISSED** [review]

**Comment by kisvegabor** [comment]
Oh, just one not: please use `circle_buf` instead of `circle_buff` for consistency. We use `buf` in other names.

**Comment by W-Mai** [comment]
> Oh, just one not: please use `circle_buf` instead of `circle_buff` for consistency. We use `buf` in other names.

That's ok, let me do it.

**Comment by W-Mai** [comment]
Done

**Comment by W-Mai** [comment]
cc @FASTSHIFT @XuNeo

**Comment by W-Mai** [comment]
@kisvegabor Is there anything else that needs to be modified? 😁

**Review by kisvegabor — DISMISSED** [review]
Looks good, thank you!

File: src/misc/lv_circle_buf.h [hunk 1]
```diff
@@ -0,0 +1,192 @@
+/**
+* @file lv_circle_buf.h
+*
+ */
+
+
+#ifndef LV_circle_buf_H
```
Comment (FASTSHIFT) [inline src/misc/lv_circle_buf.h]: LV_CIRCLE_BUF_H

File: src/misc/lv_circle_buf.h [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/*********************
+ *      INCLUDES
+ *********************/
+
+#include "lv_types.h"
+#include "lv_assert.h"
```
Comment (FASTSHIFT) [inline src/misc/lv_circle_buf.h]: move to C source file include

**Review by FASTSHIFT — CHANGES_REQUESTED** [review]

File: src/misc/lv_circle_buf.c [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    circle_buf->array = *array;
+    circle_buf->head = 0;
+    circle_buf->tail = 0;
+    circle_buf->inner_alloc = false;
+
+    circle_buf_prepare_empty(circle_buf);
+
+    return circle_buf;
+}
+lv_result_t lv_circle_buf_resize(lv_circle_buf_t * circle_buf, const uint32_t capacity)
```
Comment (FASTSHIFT) [inline src/misc/lv_circle_buf.c:117]: Add Line Break

File: src/misc/lv_circle_buf.c [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+{
+    if(lv_array_resize(&circle_buf->array, capacity) == false) {
```
Comment (FASTSHIFT) [inline src/misc/lv_circle_buf.c:121]: LV_ASSERT_NULL(circle_buf)

File: src/misc/lv_array.h [hunk 5]
```diff
@@ -51,13 +53,24 @@ struct _lv_array_t {
...
+/**
+ * Init an array from a buffer.
+ * @note The buffer must be large enough to store `capacity` elements. The array will not release the buffer and reallocate it.
+ *       The user must ensure that the buffer is valid during the lifetime of the array. And release the buffer when the array is no longer needed.
+ * @param array pointer to an `lv_array_t` variable to initialize
+ * @param buff pointer to a buffer to use as the array's data
+ * @param capacity the initial capacity of the array
+ * @param element_size the size of an element in bytes
+ */
+void lv_array_init_from_buff(lv_array_t * array, void * buff, uint32_t capacity, uint32_t element_size);
```
Comment (FASTSHIFT) [inline src/misc/lv_array.h]: lv_array_init_from_buf

File: src/misc/lv_circle_buf.h [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+lv_circle_buf_t * lv_circle_buf_create(uint32_t capacity, uint32_t element_size);
+
+/**
+ * Create a circle buffer from an existing buffer
+ * @param buff pointer to a buffer
+ * @param capacity the maximum number of elements in the buffer
+ * @param element_size the size of an element in bytes
+ * @return pointer to the created buffer
+ */
+lv_circle_buf_t * lv_circle_buf_create_from_buff(void * buff, uint32_t capacity, uint32_t element_size);
```
Comment (FASTSHIFT) [inline src/misc/lv_circle_buf.h]: buff -> buf

**Review by FASTSHIFT — APPROVED** [review]

**Review by XuNeo — APPROVED** [review]

**Pull request merged**